Thesis.org

```diff
@@ -1,11 +1,725 @@
+It is therefore important to add further abstraction to the bind logic, to allow us to implement some kind of TCO around it.
+*** TCO in ClojureScript
+While ClojureScript isn't offering TCO out-of-the-box for every function call, it does provide the ~loop~ and ~recur~ functions which do a locale rewrite of the code into a loop[14]. This allows for worry free function invocations, no matter the depth.
+
+Another way to gain TCO is to use what is called a trampoline. Instead of stepping deeper and deeper into nested function calls, the function is called once and the result, which is a function, is saved. Now for as long as the function returns another function, we will call the result. If a value is returned, we will stop and return that instead, breaking our ~loop~. Listing 18 shows a crude example. See [12] for a more detailed explanation.
+#+CAPTION: Example of trampolines in JS. ~optimized~ calls ~work~ initially and loops for as long as it returns a ~function~.
+#+begin_src javascript
+  function work(y) {
+      return (y === 0) ? true : () => work(y-1);
+  }
+
+  function optimized(x) {
+      let r = work(x);
+      while (typeof r === 'function') r = r();
+      return r;
+  }
+#+end_src
+*** Custom TCO
+While ~loop~ provides a perfectly fine way to get the benefits of TCO for synchronizing functions, in order to work with the asynchronous, action driven, approach that reacl-c uses, a custom implementation needs to be developed. We will revisit the concept of the trampoline again during our implementation.
+* Implementation
+Previously we have discussed which functions our API should offer. Now we will focus on implementing these functions and their associated features. Aside from delivering the necessary functionality, we will also talk about optimizing the ~runner~ function and implementing macros for ease of use.
+** then
+Discussed earlier, the ~then~ function needs to compose steps together. The basic idea is trivial. Take a ~Prog~ and a ~continuation~ and return a ~Bind~, which is nothing but a container ~record~ type that holds both of these values. Listing 18 shows the definition of a ~record~ type as well as the ~make-bind~ function.
+#+CAPTION: Record definition of the ~bind~ type and a helper function to filter out incorrect values.
+#+begin_src clojure
+  (defrecord Bind [prog cont])
+
+  (defn make-bind [prog cont]
+    {:pre [(prog? prog)]}
+    (->Bind prog cont))
+#+end_src
+
+However, just getting a ~Prog~ every time would be of little use. A ~Prog~ just contains a single step. Things get interesting if we want to pass a ~Bind~, because we cannot just wrap the ~Bind~ again, as it already contains a ~Prog~.
```

Comment:
Very nice section!

Reply:
I still think it would be better to say (in the very beginning of the thesis) that `bind` is not a type of its own but a value constructor for the type `prog`.